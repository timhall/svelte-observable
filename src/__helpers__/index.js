export async function load(store, wait) {
  const values = [];
  store.subscribe(value => {
    values.push(value);
  });
  await tick(wait);

  return values;
}

export function check(promise) {
  if (Array.isArray(promise)) {
    return Promise.all(promise.map(check));
  }

  return Promise.race([promise, Promise.resolve('(pending)')])
    .then(value => {
      if (value === '(pending)') return { pending: true };
      else return { fulfilled: true, value };
    })
    .catch(error => {
      return { rejected: true, error };
    });
}

export function tick(count = 0) {
  if (count === 0) {
    return new Promise(resolve => resolve());
  } else {
    return new Promise(resolve => {
      process.nextTick(resolve);
    }).then(() => (count > 1 ? tick(count - 1) : undefined));
  }
}
