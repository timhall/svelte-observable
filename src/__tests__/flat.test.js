import Observable from 'zen-observable';
import { readable } from 'svelte/store';
import { load, check, tick } from '../__helpers__';
import { flat } from '../';

const Store = {
  of(initial, ...values) {
    return readable(set => {
      (async () => {
        for (const value of values) {
          await tick();
          set(value);
        }
      })();
    }, initial);
  }
};

it('should flatten stores', async () => {
  let store = Store.of(1, 2, 3);
  let flattened = flat(store);
  let values = await load(flattened, 3);

  expect(values).toEqual([1, 2, 3]);

  store = readable(async set => {
    set(Store.of(1, 2, 3));
    await tick(1);
    set(Store.of(4, 5, 6));
    await tick(1);
    set(Store.of(7, 8, 9));
  });
  flattened = flat(store);
  values = await load(flattened, 5);

  expect(values).toEqual([1, 2, 3, 4, 5, 6, 7, 8, 9]);
});

it('should flatten observables', async () => {
  let observable = Observable.of(1, 2, 3);
  let flattened = flat(observable);

  let values = await load(flattened);
  let states = await check(values);

  expect(states[0].pending).toBe(true);
  expect(states[1].value).toEqual(1);
  expect(states[2].value).toEqual(2);
  expect(states[3].value).toEqual(3);

  observable = new Observable(async observer => {
    observer.next(Observable.of(1, 2, 3));
    await tick();
    observer.next(Observable.of(4, 5, 6));
    await tick();
    observer.next(Observable.of(7, 8, 9));
  });
  flattened = flat(observable);

  values = await load(flattened, 5);
  states = await check(values);

  expect(states).toMatchSnapshot();
});
