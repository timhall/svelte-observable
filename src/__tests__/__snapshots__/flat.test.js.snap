// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should flatten observables 1`] = `
Array [
  Object {
    "pending": true,
  },
  Object {
    "pending": true,
  },
  Object {
    "fulfilled": true,
    "value": 1,
  },
  Object {
    "fulfilled": true,
    "value": 2,
  },
  Object {
    "fulfilled": true,
    "value": 3,
  },
  Object {
    "pending": true,
  },
  Object {
    "fulfilled": true,
    "value": 4,
  },
  Object {
    "fulfilled": true,
    "value": 5,
  },
  Object {
    "fulfilled": true,
    "value": 6,
  },
  Object {
    "pending": true,
  },
  Object {
    "fulfilled": true,
    "value": 7,
  },
  Object {
    "fulfilled": true,
    "value": 8,
  },
  Object {
    "fulfilled": true,
    "value": 9,
  },
]
`;
